### DISCUSSION TITLE
Persisted store state being reset after a couple refreshes

### DISCUSSION THREAD

**Discussion by shinobu-uwu** [pmndrs/zustand#2711] [Q&A] [answered]
My stores are being reset every time after I refresh the page twice, initially my app is like this
<img width="400" alt="image" src="https://github.com/user-attachments/assets/80a9ba4e-9284-4e3f-9784-4fd9512f5252">
After a few clicks and refreshing the page
<img width="402" alt="image" src="https://github.com/user-attachments/assets/18abc1d4-486c-4630-a23b-031117d251e1">
However, if I refresh the page again it goes back to 0
<img width="405" alt="image" src="https://github.com/user-attachments/assets/bc009521-b543-4f7f-a7fe-6ab1d87aaea3">

I'm not sure what is going here, this was working and suddenly stopped with me not touching anything. Here is the code I'm using to test

```ts
// counter.ts
import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';
import { Storage } from '~/stores/storage';

export type CounterState = {
    count: number;
    loading: boolean;
    setLoading: (loading: boolean) => void;
    increment: () => void;
    decrement: () => void;
};

export const useCounterStore = create<CounterState>()(
    persist(
        (set) => ({
            count: 0,
            loading: true,
            setLoading: (loading) => set({ loading }),
            increment: () => set((state) => ({ count: state.count + 1 })),
            decrement: () => set((state) => ({ count: state.count - 1 })),
        }),
        {
            name: 'counter',
            storage: createJSONStorage(() => Storage),
            onRehydrateStorage: (state) => state.setLoading(false),
        }
    )
);
```
```tsx
// index.tsx
import { createFileRoute } from '@tanstack/react-router';
import { useCounterStore } from '~/stores/counter';

export const Route = createFileRoute('/')({
    component: Index,
});

function Index() {
    const { loading, count, increment } = useCounterStore();

    if (loading) {
        return <span>Loading...</span>;
    }

    debugger;
    return (
        <div style={{ padding: 10 }}>
            <span>{count}</span>
            <br />
            <button onClick={increment}>Login</button>
        </div>
    );
}
```
```ts
// storage.ts
import type { StateStorage } from 'zustand/middleware';
import { SecureStorage } from '@plasmohq/storage/secure';

export const Storage: StateStorage = {
    getItem: async (name: string): Promise<string | null> => {
        const storage = new SecureStorage();
        await storage.setPassword('password');
        const item = await storage.getItem(name);

        return item === undefined ? null : item;
    },
    setItem: async (name: string, value: string): Promise<void> => {
        const storage = new SecureStorage();
        await storage.setPassword('password');

        return storage.setItem(name, value);
    },
    removeItem: async (name: string): Promise<void> => {
        const storage = new SecureStorage();
        await storage.setPassword('password');

        return storage.removeItem(name);
    },
};
```

I checked the value of the key in the secure storage during app startup and it was present, so looks like zustand is the one resetting the value

**Comment by dbritto-dev** [comment]
@shinobu-uwu we need more context here: SPA or MPA? if is MPA, is it Next.js or something else?

**Reply by shinobu-uwu** [reply]
@dbritto-dev It's an SPA, I am creating a chrome extension, even though it shouldn't matter, I'm using [plasmo](https://www.plasmo.com/), secure storage is just encrypted `storage.sync`, however if I change to use `localStorage` the behavior is still the same

**Reply by dbritto-dev** [reply]
@shinobu-uwu it should matter, the usage of Zustand is different between SPA and MPA. BTW, would you mind creating a minimal repro on stactkblitz?

**Reply by shinobu-uwu** [reply]
Oops, i meant it shouldn't matter if it's a chrome extension or not, because it's an SPA anyway. Sure I'll do it soon

**Reply by shinobu-uwu** [reply]
🤔 plasmo won't work on stackblitz since it's not an extension, however I've used the default `localStorage` here and the behavior is weird, `count` is undefined when that shouldn't be possible

https://stackblitz.com/edit/github-rdp8de?file=src%2Fstore.ts

**Reply by shinobu-uwu** [reply]
If you're willing to go a little further I created [this repo](https://github.com/shinobu-uwu/zustand-repro) that the exact issue occurs, the step-by-step to load the chrome extension can be found [here](https://docs.plasmo.com/framework#loading-the-extension-in-chrome)

**Reply by dbritto-dev** [reply, marked as answer]
@shinobu-uwu here you go

```diff
- onRehydrateStorage: (state) => state.setLoading(false),
+ onRehydrateStorage: (state) => {
+  return () => void state.setLoading(false)
+ },
```

**Reply by shinobu-uwu** [reply]
Holy moly, thanks so much! I was nearly losing my head with this, I kept reading the documentation page but I let that pass 😅

**Reply by dbritto-dev** [reply]
@shinobu-uwu yeah, we are working on revamping docs to avoid that haha

**Reply by dbritto-dev** [reply]
for now apis, hooks and then I'll push docs for middlewares too

**Reply by dbritto-dev** [reply]
everything with working examples
